### DISCUSSION TITLE
Implicit member type declaration

### DISCUSSION THREAD

**Discussion by taori** [dotnet/csharplang#6581] [Language Ideas]
This idea may be discarded because of how expensive it might be implementation wise but here goes nothing:

I often find myself working with tuples when i want to do some simple processing of data, aggregating and whatnot, then realize that parameter maintenance becomes a pain and i have to refactor all my code in order to pass parameters in a sane fashion once the call hierarchy exceeds a threshold. Let's consider the following scenario: (ignore data structure, just making some larger type declarations)

```
public (string aggregateId, Dictionary<int, Dictionary<string, string>> data) GetSomeData(string someKey)
public void ProcessData((string aggregateId, Dictionary<int, Dictionary<string, string>> data))
public void ProcessDataSomeMore((string aggregateId, Dictionary<int, Dictionary<string, string>> data))
```

### implementation of GetSomeData:

```
public (string aggregateId, Dictionary<int, Dictionary<string, string>> data) GetSomeData(string someKey)
{
    return (aggregateId: "hi", data: new Dictionary<int, Dictionary<string,string>>());
}
```

If the signature GetSomeData changes i usually make my method return void, then let IDE generate the return type. then i have to copy that type wherever it is used in call chain.

Mind you i know how this could be solved clean now: `record(string AggregateId, Dictionary<string, string>> Data)` but i don't want to keep updating my code all the time so here is what i propose:

### syntax using proposal:
```
public var DataHolder GetSomeData(string someKey)
{
    return (aggregateId: "hi", data: new Dictionary<int, Dictionary<string,string>>());
}
void ProcessData(DataHolder dataHolder)
void ProcessDataSomeMore(DataHolder dataHolder)
```

### created by compiler:
```
// implicitly generated from returns within a method: 
record DataHolder(string AggregateId, Dictionary<string, string>> Data)
```

Not sure if this is possible, but it sure would be extremely convenient to work with.

**Comment by HaloFour** [comment]
See: https://github.com/dotnet/csharplang/discussions/6452

**Reply by taori** [reply]
Was unware of that proposal. I'll check out if it is any different from mine

**Reply by taori** [reply]
@HaloFour As far as i can tell that idea is similar, but appears, from what i can tell, to have one significant difference: type declaration happening somewhere else and inference potentially being a deep process, no? In my case there is no deep inspection required to infer the actual type. It would happen in the body of the method.

I have also read about being explicit about a methods contract which i understand, but it assumes awareness of this being a concern when developing a library - You either are aware and care, or you don't and you break stuff. Having to specify the type in the method does not protect an api consumer from an api developer doing that - the only thing protecting you from that is not the language constraints, but that the api developer knows about not breaking public api for that reason, or indicating breaking api by elevating the major version of a library.

Apart from this, if this proposal was also restricted to private/internal/file methods it would make the contractual part of the argumentation void, no? Whenever i use nonpublic identifiers my mindset is that it is none of the conumers business how i process data, to transform it into something the consumer may want it to wortk with. Any methods i create just help me encapuslate the complexity into smaller portions - having to be explicit with the types (which is not fun when the type becomes 10+ chars) adds no value to the consumer of a library or its producr, but only to the performance of roslyn.

If my proposal was too heavy on perf despite not allowing nested var inference, but generating the type inplace i would think that argumentation is perfectly reasonable, but if in place generating a type would not really be an issue i think this would add great value for a significant amount of developers.

**Reply by HaloFour** [reply]
There is no practical difference, if the return type is `var` then the compiler needs to do the inference.  As methods can call other methods, and those other methods have inferred return types, you end up having to recompile/reinterpret a large amount of code in order to determine the inferred return type.  That is _extremely_ expensive and would essentially break a lot of the shortcuts that the compiler relies on to provide realtime analysis.

**Reply by taori** [reply]
Hm, i guess unless there is a very important reason for roslyn to do more recompile passes this feature won't happen then. bummer

**Comment by TahirAhmadov** [comment]
I could see something different:
```cs
public (record DataHolder(string aggregateId, Dictionary<int, Dictionary<string,string>> data)) GetSomeData(string someKey)
{
  return new('', new());
}
void ProcessData(DataHolder dataHolder);
void ProcessDataSomeMore(DataHolder dataHolder);
```
But still, this seems to be **much** more trouble than it's worth.

**Reply by rbwhitaker** [reply]
An interesting variation. An inline record definition. But I'd agree it seems a lot for what you get. I'm not in the habit of looking for entire type definitions--especially not ones used across the program--in a return type. And just defining the record on its own really isn't bad.
